new-log-viewer/src/services/decoders/JsonlDecoder.ts

```diff
@@ -0,0 +1,240 @@
+        for (const propName of this.#propertyNames) {
+            if (propName in logEvent) {
+                const placeholder = `%${propName}`;
+                const propValue = logEvent[propName];
+                let propValueString: string;
+                if ("undefined" === typeof propValue || null === propValue) {
+                    console.error(`Unexpected undefined logEvent[${propName}]`);
+                    propValueString = "undefined";
+                } else if ("object" === typeof propValue) {
+                    propValueString = JSON.stringify(propValue);
+                } else {
+                    propValueString = propValue.toString();
+                }
+                input = input.replace(placeholder, propValueString);
+            }
+        }
+
+        input += "\n";
+
+        return input;
+    }
+
+    /**
+     * Extracts the verbosity from the given log event.
+     *
+     * @param logEvent The log event containing the verbosity.
+     * @return The extracted verbosity.
+     */
+    #extractVerbosity (logEvent: JsonObject) {
+        let verbosity = LOG_VERBOSITY.NONE;
+        const verbosityString: string = logEvent[this.#verbosityPropName] as string;
+        if (false === (verbosityString in LOG_VERBOSITY)) {
+            console.error(`Unable to find verbosity from key ${this.#verbosityPropName}` +
+                ` of type ${typeof verbosityString}`);
+        } else {
+            verbosity = LOG_VERBOSITY[verbosityString as (keyof typeof LOG_VERBOSITY)];
+        }
+
+        return verbosity;
+    }
+
+    constructor (dataArray: Uint8Array) {
+        this.#dataArray = dataArray;
+    }
+
+    setDecodeOptions (options: JsonlDecodeOptionsType): boolean {
+        this.#setTextPattern(
+            options.textPattern
+        );
+        this.#timestampPropName = options.timestampPropName;
+        this.#verbosityPropName = options.verbosityPropName;
+
+        return true;
+    }
+
+    buildIdx (): number {
+        const text = JsonlDecoder.TEXT_DECODER.decode(this.#dataArray);
+        const split = text.split("\n");
+        for (const line of split) {
+            if (0 === line.length) {
+                continue;
+            }
+            try {
+                const logEvent = JSON.parse(line) as JsonObject;
+                this.#logEvents.push(logEvent);
+            } catch (e) {
+                if (e instanceof SyntaxError) {
+                    console.error(e, line);
+                }
+            }
+        }
+
+        return this.#logEvents.length;
+    }
+
+    decode (results: DecodeResultType[], startIdx: number, endIdx: number): boolean {
+        console.assert(0 === results.length, "results array is non-empty");
+
+        for (let logEventIdx = startIdx; logEventIdx < endIdx; logEventIdx++) {
+            const logEvent = this.#logEvents[logEventIdx];
+            if ("undefined" === typeof logEvent) {
+                return false;
```

Comment:
> How could this happen? Are you anticipating an OOB error?

I don't anticipate this to happen although the subscripting operator does return "undefined" when such happens. 

> If so, couldn't we check the range is within the bounds of #logEvents?

Unfortunately, TypeScript hasn't been smart enough to assume the non-`undefined` type even with explicit range checking. 

(With TypeScript not being smart, imagine how much overhead it would introduce to implement an algorithm like Binary Search, lol.)

Reply:
Hm, in that case, would using a slice avoid typescript's complaints?